src/interface.tsx

```diff
@@ -0,0 +1,23 @@
+export interface ITodoItem {
+  text: string;
+}
+
+export interface ITodoItemProps {
+  type: "todo" | "done";
+  todo: string;
+  idx: number;
+  deleteCurrentList: (idx: number) => void;
+  addToggleList: (text: string) => void;
+}
+
+export interface IState {
+  todoList: Array<ITodoItem>;
+  doneList: Array<ITodoItem>;
+  dispatch: React.Dispatch<IAction>;
+}
+
+export interface IAction {
+  type: "ADD_TODO" | "DELETE_TODO" | "ADD_DONE" | "DELETE_DONE";
+  todo?: { text: string };
+  idx?: number;
+}
```

Comment:
action은 명확하게 나눠줘야 합니다. 현재 코드에서는 `ADD_DONE` 타입의 action에 payload로 `idx`를 넘겨준다고 해도 컴파일타임에 에러가 없습니다. (하지만 명백히 잘못된 payload 전달이죠) 타입스크립트의 목적중 하나가 런타임에 에러를 방지하려는 것임을 고려해보면 꼭 지켜져야 하는 부분입니다.

```suggestion
export interface IAction {
  type: "ADD_TODO";
  todo: { text: string };
} |
{
  type: "DELETE_TODO";
  idx: number;
} |
{
  type: "ADD_DONE";
  todo: { text: string };
} |
{
  type: "DELETE_DONE";
  idx: number;
}
```

Reply:
action에 따라 다른 payload를 받는 것을 어떻게 선언해줘야 할 지 몰라서 저런 애매한 코드를 작성해버렸네요.. 말씀해주신대로 명확하게 작성하도록 하겠습니다. 감사합니다!